soc/arm/silabs_sim3/sim3u/soc.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/irq.h>
+
+#include <SI32_CLKCTRL_A_Type.h>
+#include <SI32_PBCFG_A_Type.h>
+#include <SI32_PBSTD_A_Type.h>
+#include <SI32_RSTSRC_A_Type.h>
+#include <SI32_VMON_A_Type.h>
+#include <SI32_WDTIMER_A_Type.h>
+#include <si32_device.h>
+#include <silabs_crossbar_config.h>
+
+static void gpio_init(void)
+{
+	/* Enable clocks that we need in any case */
+	SI32_CLKCTRL_A_enable_apb_to_modules_0(
+		SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG0_PB0 | SI32_CLKCTRL_A_APBCLKG0_FLASHCTRL0);
+
+	/* Enable misc clocks.
+	 * We may or may nor need all of that but it includes some basics like
+	 * LOCK0, WDTIMER0 and DMA. Doing it here prevents the actual drivers
+	 * needing to know whoelse needs one of these clocks.
+	 */
+	SI32_CLKCTRL_A_enable_apb_to_modules_1(
+		SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG1_MISC0 | SI32_CLKCTRL_A_APBCLKG1_MISC1);
+	SI32_CLKCTRL_A_enable_ahb_to_dma_controller(SI32_CLKCTRL_0);
+
+	/* Apply pinctrl gpio settings */
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_0, PORTBANK_0_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_0, PORTBANK_0_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_0, PORTBANK_0_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_0, PORTBANK_0_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_0, PORTBANK_0_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_1, PORTBANK_1_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_1, PORTBANK_1_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_1, PORTBANK_1_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_1, PORTBANK_1_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_1, PORTBANK_1_PINS_ANALOG);
+
+	SI32_PBSTD_A_write_pins_high(SI32_PBSTD_2, PORTBANK_2_PINS_HIGH);
+	SI32_PBSTD_A_write_pins_low(SI32_PBSTD_2, PORTBANK_2_PINS_LOW);
+	SI32_PBSTD_A_set_pins_digital_input(SI32_PBSTD_2, PORTBANK_2_PINS_DIGITAL_INPUT);
+	SI32_PBSTD_A_set_pins_push_pull_output(SI32_PBSTD_2, PORTBANK_2_PINS_PUSH_PULL_OUTPUT);
+	SI32_PBSTD_A_set_pins_analog(SI32_PBSTD_2, PORTBANK_2_PINS_ANALOG);
```

Comment:
-1 on custom pin control mechanisms.

Reply:
can you expand on what you mean by "custom"? Are you referring to https://github.com/zephyrproject-rtos/zephyr/pull/64209#discussion_r1368481084 ? Any comments on my reply to that?